### DISCUSSION TITLE
`std::unreachable` doing nothing, currently

### DISCUSSION THREAD

**Discussion by AlexBAV** [microsoft/STL#3051] [Ideas] [answered]
It looks like with the current compiler implementation, `std::unreachable` does not affect optimizer. `__assume(false)`, on the other hand, seems to do the job. 

`std::unreachable` is implemented by means of `__assume(false)`, however, as illustrated in the following issue:

https://developercommunity.visualstudio.com/t/Suboptimal-code-generation-for-std::unre/10071879
https://godbolt.org/z/58T55q7Mf

compiler misses the opportunity in this case.

According to the compiler team's feedback, they are currently not very interested in fixing this. However, it looks like if `__forceinline` is added to the function declaration, compiler is again capable of using the optimization hint, at least in simple tests:

https://github.com/microsoft/STL/blob/febb6430ac48a6650b174d567a335afcc0db8dc8/stl/inc/utility#L816

**Comment by AlexBAV** [comment, marked as answer]
The following illustrates this (`foo` uses `__assume(false)`, `bar` uses `std::unreachable` and `good_bar` uses patched `unreachable`):

https://godbolt.org/z/a9d33hc5b

**Reply by StephanTLavavej** [reply]
Thanks, this codegen helps me confidently approve #3055.
